File: src/ClaudeRecall.Hook/Projects.cs
// Projects.cs — port of src/claude_recall/projects.py.
//
// Derives the Claude Code session-archive slug from a working directory.
// Must produce byte-identical output to the Python implementation so the
// binary and Python CLI resolve the same project_slug for the same cwd.

using Microsoft.Data.Sqlite;

namespace ClaudeRecall.Hook;

internal static class Projects
{
    /// <summary>
    /// Derive the Claude Code project slug from an absolute directory path.
    /// Windows: "E:\\Documents\\Work" → "e--Documents-Work"
    /// Unix:    "/Users/m/dev"       → "-Users-m-dev"
    /// Pure function — no filesystem access, no DB.
    /// </summary>
    public static string SlugFromPath(string path)
    {
        if (string.IsNullOrEmpty(path))
        {
            return string.Empty;
        }

        // Lowercase the Windows drive letter only (preserve rest of the case).
        if (path.Length >= 2 && path[1] == ':' && char.IsLetter(path[0]))
        {
            path = char.ToLowerInvariant(path[0]) + path[1..];
        }

        // Replace ':' then '\' then '/' — order matches the Python sequence.
        var chars = path.ToCharArray();
        for (int i = 0; i < chars.Length; i++)
        {
            if (chars[i] == ':' || chars[i] == '\\' || chars[i] == '/')
            {
                chars[i] = '-';
            }
        }
        return new string(chars);
    }

    /// <summary>
    /// Resolve the project_slug stored in the index that corresponds to ``cwd``.
    /// Returns the exact stored slug (case-preserved) when a case-insensitive
    /// match exists, else the canonical lowercase derivation, else null.
    /// Mirrors ``projects.resolve_project_slug`` in Python.
    /// </summary>
    public static string? ResolveProjectSlug(SqliteConnection conn, string? cwd)
    {
        cwd ??= Environment.CurrentDirectory;
        var canonical = SlugFromPath(Path.GetFullPath(cwd));
        if (string.IsNullOrEmpty(canonical))
        {
            return null;
        }

        using var cmd = conn.CreateCommand();
        cmd.CommandText = "SELECT project_slug FROM sessions " +
                          "WHERE LOWER(project_slug) = LOWER($s) LIMIT 1";
        cmd.Parameters.AddWithValue("$s", canonical);
        var result = cmd.ExecuteScalar();
        return result is string s ? s : canonical;
    }
}

File: src/claude_recall/projects.py
"""Project-slug helpers (v0.2.1).

Claude Code stores session archives under
``~/.claude/projects/<project-slug>/<uuid>.jsonl`` where ``<project-slug>`` is
derived from the working directory Claude Code was launched in. Historically
the convention has drifted between ``E--...`` and ``e--...`` forms; the current
convention is lowercase drive letter. This module computes the canonical slug
from a path and, when a DB connection is supplied, resolves it against the
sessions table case-insensitively so both conventions work.
"""

from __future__ import annotations

import sqlite3
from pathlib import Path


def slug_from_path(path: Path | str) -> str:
    """Derive the Claude Code project slug from an absolute directory path.

    Rules (matching the Claude Code convention observed in real archives):
    - Drive letter on Windows is lowercased (``E:\\Documents\\...`` → ``e--Documents-...``)
    - ``:`` becomes ``-``
    - ``\\`` and ``/`` each become ``-``
    - The path is not URL- or percent-encoded; unusual characters pass through

    This is a pure function. No filesystem access. No DB lookup.
    """
    s = str(path)
    if len(s) >= 2 and s[1] == ":" and s[0].isalpha():
        s = s[0].lower() + s[1:]
    return s.replace(":", "-").replace("\\", "-").replace("/", "-")


def resolve_project_slug(
    conn: sqlite3.Connection, cwd: Path | str | None = None
) -> str | None:
    """Resolve the project_slug that corresponds to ``cwd`` in the index.

    Returns the exact slug stored in ``sessions.project_slug`` if a case-
    insensitive match exists, else the canonical (lowercase) slug derived
    from the path, else None if cwd is empty.

    The DB lookup lets auto-scoping work for users whose archive still uses
    the older ``E--`` uppercase convention — we return the actual stored
    slug so the ``WHERE project_slug = ?`` filter in the search SQL hits.
    """
    if cwd is None:
        cwd = Path.cwd()
    canonical = slug_from_path(Path(cwd).absolute())
    if not canonical:
        return None

    row = conn.execute(
        "SELECT project_slug FROM sessions "
        "WHERE LOWER(project_slug) = LOWER(?) LIMIT 1",
        (canonical,),
    ).fetchone()
    if row is not None:
        return row["project_slug"]
    return canonical
